[8.0.1.0.0][partner_workflow_policy]

azmimr67 opened the pull request:
Partner workflow policy
Policy untuk mark as customer dan supplier
Untuk field yang di read only masih di mark di res_partner.py baris 46 sd 87

azmimr67 committed:
[8.0.1.0.0][partner_workflow_policy]
Partner workflow policy
Policy untuk mark as customer dan supplier
Untuk field yang di read only masih di mark di res_partner.py baris 46 sd 87

mikevhe18 commented on partner_workflow_policy/models/res_partner.py:
```diff
@@ -0,0 +1,204 @@
+# Copyright 2020 PT Simetri Sinergi Indonesia
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from openerp import models, fields, api, _
```
Import "_(_Undescore_)" dihapus om karena tidak terpakai

azmimr67 replied:
done

mikevhe18 commented on partner_workflow_policy/__openerp__.py:
```diff
@@ -0,0 +1,22 @@
+# -*- coding: utf-8 -*-
+# Copyright 2020 OpenSynergy Indonesia
```
Tambahkan # Copyright 2020 PT Simetri Sinergi Indonesia

azmimr67 replied:
done

mikevhe18 commented on partner_workflow_policy/models/res_partner.py:
```diff
@@ -0,0 +1,204 @@
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from openerp import models, fields, api, _
+from openerp.exceptions import Warning as UserError
```
Dihapus Om karena tidak terpakai

azmimr67 replied:
done

mikevhe18 commented on partner_workflow_policy/models/res_partner.py:
```diff
@@ -0,0 +1,204 @@
+    #     states={
+    #         "draft": [
+    #             ("readonly", False),
+    #         ],
+    #     },
+    # )
+    # user_id = fields.Many2one(
+    #     comodel_name="res.users",
+    #     readonly=True,
+    #     states={
+    #         "draft": [
+    #             ("readonly", False),
+    #         ],
+    #     },
+    # )
+    # category_id = fields.Many2many(
+    #     comodel_name="res.partner.category",
+    #     readonly=True,
+    #     states={
+    #         "draft": [
+    #             ("readonly", False),
+    #         ],
+    #     },
+    # )
+    # ref = fields.Char(
+    #     readonly=True,
+    #     states={
+    #         "draft": [
+    #             ("readonly", False),
+    #         ],
+    #     },
+    # )
+    # lang = fields.Selection(
+    #     readonly=True,
+    #     states={
+    #         "draft": [
+    #             ("readonly", False),
+    #         ],
+    #     },
+    # )
```
Untuk field-field yang butuh di-readonly, dilakuinnya di view aja om pake attributes "attrs". Soalnya nanti di res.users akan jadi readonly juga dan membuat error

mikevhe18 commented on partner_workflow_policy/models/res_partner.py:
```diff
@@ -0,0 +1,204 @@
+    # Policy Fields
+    confirm_ok = fields.Boolean(
+        string="Can Confirm",
+        compute="_compute_policy"
```
Tambahkan ",(koma)" diakhir

azmimr67 replied:
done

mikevhe18 commented on partner_workflow_policy/models/res_partner.py:
```diff
@@ -0,0 +1,204 @@
+    )
+    valid_ok = fields.Boolean(
+        string="Can Valid",
+        compute="_compute_policy"
```
Tambahkan ",(koma)" diakhir

azmimr67 replied:
done

mikevhe18 commented on partner_workflow_policy/models/res_partner.py:
```diff
@@ -0,0 +1,204 @@
+    )
+    restart_ok = fields.Boolean(
+        string="Can Restart",
+        compute="_compute_policy"
```
Tambahkan ",(koma)" diakhir

azmimr67 replied:
done

mikevhe18 commented on partner_workflow_policy/models/res_partner.py:
```diff
@@ -0,0 +1,204 @@
+    # Policy Fields untuk customer
+    mark_customer_ok = fields.Boolean(
+        string="Can Mark As Customer",
+        compute="_compute_policy"
```
Tambahkan ",(koma)" diakhir

azmimr67 replied:
done

mikevhe18 commented on partner_workflow_policy/models/res_partner.py:
```diff
@@ -0,0 +1,204 @@
+    )
+    unmark_customer_ok = fields.Boolean(
+        string="Can Un Mark As Customer",
+        compute="_compute_policy"
```
Tambahkan ",(koma)" diakhir

azmimr67 replied:
done

mikevhe18 commented on partner_workflow_policy/models/res_partner.py:
```diff
@@ -0,0 +1,204 @@
+    # Policy Fields untuk customer
+    mark_supplier_ok = fields.Boolean(
+        string="Can Mark As Supplier",
+        compute="_compute_policy"
```
Tambahkan ",(koma)" diakhir

azmimr67 replied:
done

mikevhe18 commented on partner_workflow_policy/models/res_partner.py:
```diff
@@ -0,0 +1,204 @@
+    )
+    unmark_supplier_ok = fields.Boolean(
+        string="Can Un Mark As Supplier",
+        compute="_compute_policy"
```
Tambahkan ",(koma)" diakhir

azmimr67 replied:
done

mikevhe18 commented on partner_workflow_policy/views/res_partner_views.xml:
```diff
@@ -0,0 +1,87 @@
+            <field name="inherit_id" ref="base.view_partner_form"/>
+            <field name="arch" type="xml">
+            <data>
+                <xpath expr="//page[@name='sales_purchases']" position="after">
+                    <page name="policy" string="Policy" groups="base.group_system">
+                        <group name="policy_1" colspan="4" col="2">
+                            <field name="confirm_ok"/>
+                            <field name="valid_ok"/>
+                            <field name="restart_ok"/>
+                            <field name="mark_customer_ok"/>
+                            <field name="unmark_customer_ok"/>
+                            <field name="mark_supplier_ok"/>
+                            <field name="unmark_supplier_ok"/>
+                        </group>
+                    </page>
+                    <page name="log" string="Log">
+                        <group name="group_log_1" colspan="4" col="3">
+                            <field name="confirm_date" string="Confirm"/>
+                            <field name="confirm_user_id" nolabel="1"/>
+                            <field name="valid_date" string="Validate"/>
+                            <field name="valid_user_id" nolabel="1"/>
+                            <field name="restart_date" string="Restart"/>
+                            <field name="restart_user_id" nolabel="1"/>
+                        </group>
+                    </page>
+                </xpath>
```
ditaro diatas aja om disatukan sama penambahan button

azmimr67 replied:
Jika dipindahkan ke atas posisinya tidak bisa di sebelah sales_purchase.
Jadi saya biarkan tetap disini saja

mikevhe18 commented on partner_workflow_policy/views/res_partner_views.xml:
```diff
@@ -0,0 +1,87 @@
+                    <button name="action_valid" type="object" string="Valid" class="oe_highlight" attrs="{'invisible':['|',('state','!=','confirm'),('valid_ok','=',False)]}" confirm="Validate data. Are you sure?"/>
+                    <button name="action_restart" type="object" string="Set to Draft" attrs="{'invisible':['|',('state','=','draft'),('restart_ok','=',False)]}" confirm="Restart data. Are you sure?"/>
+                    <field name="state" string="State" widget="statusbar" statusbar_visible="draft,confirm,valid"/>
+                </xpath>
```
Tambahkan field-field yang mau di-readonly menggunakan attributes.
Cth:
```
<xpath expr="//field[@name='child_ids']" position="attributes">
    <attribute name="attrs">
        {'readonly':[('state','!=', 'draft')]}
    </attribute>
</xpath>
```

azmimr67 replied:
done

azmimr67 committed:
[8.0.1.0.0][partner_workflow_policy]
Revisi menambahkan disable pada view untuk field yang tidak bloeh diubah
Revisi sesuai comment di github
